Google.Api.Gax.Grpc/GaxGrpcServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,39 @@
+﻿/*
+ * Copyright 2022 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Google.Api.Gax.Grpc;
+using Grpc.Net.Client;
+using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Extension methods for dependency injection.
+    /// </summary>
+    public static class GaxGrpcServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Adds a singleton <see cref="GrpcNetClientAdapter"/> to the given service collection,
+        /// using the default with any additional options configured via <paramref name="optionsConfigurer"/>.
+        /// </summary>
+        /// <param name="services"></param>
+        /// <param name="optionsConfigurer"></param>
+        /// <returns></returns>
+        public static IServiceCollection AddGrpcNetClientAdapter(this IServiceCollection services, Action<IServiceProvider, global::Grpc.Net.Client.GrpcChannelOptions> optionsConfigurer) =>
+            services.AddSingleton<GrpcAdapter>(provider => GrpcNetClientAdapter.Default.WithAdditionalOptions(options => optionsConfigurer(provider, options)));
+
+        /// <summary>
+        /// Adds a singleton <see cref="GrpcNetClientAdapter"/> to the given service collection,
+        /// such that any <see cref="GrpcChannel"/> created uses the service provider from
+        /// created this service collection. This enables logging, for example.
+        /// </summary>
+        /// <param name="services"></param>
+        /// <returns></returns>
+        public static IServiceCollection AddGrpcNetClientAdapter(this IServiceCollection services) =>
```

Comment:
For solving the REST Adapter bits and following with the factory idea (restricting visibility as much as possible). Also, quickly here but I'm happy to push a PR for this.

```csharp
public interface IAdapterFactory { GrpcAdapter Get(GrpcAdapterOptions options); }

public class GrpcAdapterOptions
{ 
    public IEnumerable<FileDescriptors> Descriptors { get; set; }
    public SupportedFormat SupportedFormats { get; set; }

    [Flags] // Possibly better names here.
    public enum SupportedFormat { GRpc, Json}
}

// And a similar one for Grpc.Core if we wanted to.
public class GrpcNetClientAdapterFactory : IAdapterFactory
{
    private readonly GrcpNetClientAdapter _instance;

    public GrpcNetClientAdapterFactory(IServiceProvider provider) =>
        _instance = GrcpNetClientAdapter.Default.WithAdditionalOptions(options => options.ServiceProvider = provider);

    public GrpcAdapter Get(GrcpAdapterOptions options) => 
         options.SupportedFormats.HasFlag(SupportedFormat.Grpc) ? _instance : null;
}

public class RestGrpcAdapterFactory : IAdapterFactory
{
    private readonly ConcurrentDictionary<GrpcAdapterOptions, RestGrpcAdapter> _cache = new ....;
    
    // No need for explicit constructor.

    public GrpcAdapter Get(GrpAdapterOptions options) =>
        // if the supported format includes JSON, find or add the adapter on the cache using the descriptor, else return null.
}
```

On the extension methods we always inject the factories instead of the adapters.
On the client builders we look for the factories and always pass the descriptors (which we know, even if we have to add a new abstract property) and the formats we can support (i.e. Compute only supports JSON, gapics only support gRPC for now but will support JSON at some point). If we don't get a factory or an adapter, that's fine, we don't set it and default will be used.

Reply:
Maybe. I think we can defer that part a little... I'm slightly nervous about having too many abstractions. We could potentially make the single GrpcAdapter method that's used to create a ChannelBase accept the additional option of the file descriptors, and just ignore them for other adapters. Not sure...